theano/gpuarray/dnn.py

```diff
@@ -2985,6 +3056,73 @@ def local_gpua_logsoftmax_to_dnn(op, ctx_name, inputs, outputs):
     return [out.dimshuffle(0, 1)]
 
 
+@register_opt('cudnn', 'fast_compile')
+@op_lifter([SoftmaxGrad])
+@register_opt2([SoftmaxGrad], 'cudnn', 'fast_compile')
+def local_gpua_softmax_dnn_grad(op, ctx_name, inputs, outputs):
+    if not dnn_available(ctx_name):
+        return
+    ins = []
+    for n in inputs:
+        n = as_gpuarray_variable(n, ctx_name)
+        if n.ndim != 2:
+            return
+        ins.append(n.dimshuffle(0, 'x', 1, 'x'))
+
+    out = GpuDnnSoftmaxGrad('accurate', 'instance')(
+        gpu_contiguous(ins[0]), gpu_contiguous(ins[1]))
+    return [out.dimshuffle(0, 2)]
+
+
+@register_opt('cudnn')
+@local_optimizer([GpuCAReduceCuda])
+def local_dnn_reduction(node):
+    if not isinstance(node.op, GpuCAReduceCuda):
+        return
+
+    if not dnn_available(node.inputs[0].type.context_name):
+        return
+
+    if version(raises=False) < 6000:
+        return
+
+    if node.inputs[0].dtype != node.outputs[0].dtype:
+        # We can mix float16 and float32, but not float64.
+        if (node.inputs[0].dtype == 'float64' or
+                node.outputs[0].dtype == 'float64'):
+            return
+        if node.op.acc_dtype != 'float32':
+            return
+
+    if node.inputs[0].dtype not in ['float16', 'float32', 'float64']:
+        return
+
+    if (node.inputs[0].dtype == 'float64' and
+            node.op.acc_dtype != 'float64'):
+        return
+
+    if (node.inputs[0].dtype == 'float32' and
+            node.op.acc_dtype != 'float32'):
+        return
```

Comment:
I think it means this will almost never be used.
Currently, the default `acc_dtype` for `float32` inputs on CPU is `float64`. In the old back-end, the accumulation dtype was always `float32`, because the reduction structure made it more precise.
However, now the `acc_dtype` for `GpuCAReduceCuda` set in the lifter is the CPU one, which is `float64` by default.

If it is indeed a limitation of cuDNN, we need to do one of the following:
- force `acc_dtype=float32` in `GpuCAReduceCuda` (which would be consistent with the old back-end)
- force `acc_dtype=float32` in `GpuDnnReduction`